libbfddp/bfddp_extra.c

```diff
@@ -0,0 +1,858 @@
+			  | (bs->bs_cbit << 3u)	  /* Control Plane I. bit */
+			  | (0 << 2u)		  /* Authentication bit. */
+			  | (bs->bs_demand << 1u) /* Demand mode bit */
+			  | (0 << 0u));		  /* Multi point bit. */
+
+	bcp->detection_multiplier = bs->bs_dmultiplier;
+	bcp->length = sizeof(*bcp);
+	bcp->local_id = htonl(bs->bs_lid);
+	bcp->remote_id = htonl(bs->bs_rid);
+
+	switch (bs->bs_state) {
+	case STATE_ADMINDOWN:
+	case STATE_DOWN:
+		bcp->desired_tx = htonl(bs->bs_cur_tx);
+		bcp->required_rx = htonl(bs->bs_cur_rx);
+		bcp->required_echo_rx = htonl(bs->bs_cur_erx);
+		break;
+	case STATE_INIT:
+	case STATE_UP:
+		bcp->desired_tx = htonl(bs->bs_tx);
+		bcp->required_rx = htonl(bs->bs_rx);
+		bcp->required_echo_rx = htonl(bs->bs_erx);
+		break;
+	}
+}
+
+ssize_t
+bfddp_send_control_packet(struct bfd_session *bs, void *arg)
+{
+	struct bfddp_control_packet bcp;
+	ssize_t rv;
+
+	bfddp_fill_control_packet(bs, &bcp);
+
+	rv = bfddp_callbacks.bc_tx_control(bs, arg, &bcp);
+
+	/* Update session output data. */
+	if (rv > 0) {
+		bs->bs_ctx_bytes += (size_t)rv;
+		bs->bs_ctx_packets++;
+	}
+
+	return rv;
+}
+
+void
+bfddp_session_rx_timeout(struct bfd_session *bs, void *arg)
+{
+	enum bfd_state_value pstate = bs->bs_state;
+
+	bfddp_session_reset_remote(bs);
+
+	/* Tell FRR's BFD daemon the session is down. */
+	bs->bs_state = STATE_DOWN;
+	bs->bs_diag = bs->bs_rdiag = DIAG_CONTROL_EXPIRED;
+	bfddp_session_set_slowstart(bs);
+
+	/* Disable timers if configured for passive mode. */
+	if (bs->bs_passive) {
+		bfddp_callbacks.bc_rx_control_stop(bs, arg);
+		bfddp_callbacks.bc_tx_control_stop(bs, arg);
+	}
+
+	/*
+	 * Only send notification if state changed.
+	 *
+	 * This prevents a extra notification in case the remote peer
+	 * asked and ceased to send control packets earlier than expiration
+	 * timer.
+	 */
+	if (bs->bs_state != pstate)
+		bfddp_send_session_state_change(bs);
+
+	/* Notify application about state change. */
+	bfddp_callbacks.bc_state_change(bs, arg, pstate, bs->bs_state);
+}
+
+static void
+bfddp_session_sm_admindown(__attribute__((unused)) struct bfd_session *bs,
+			   __attribute__((unused)) void *arg,
+			   __attribute__((unused)) enum bfd_state_value nstate)
+{
+	/* NOTHING. */
+}
+
+static void
+bfddp_session_sm_down(struct bfd_session *bs, void *arg,
+		      enum bfd_state_value nstate)
+{
+	switch (nstate) {
+	case STATE_ADMINDOWN:
+		/* NOTHING. */
+		break;
+	case STATE_DOWN:
+		bs->bs_state = STATE_INIT;
+
+		/* Notify state change. */
+		bfddp_send_session_state_change(bs);
+
+		/*
+		 * If passive mode we must manually send packet (TX timer is
+		 * disabled).
+		 */
+		if (bs->bs_passive)
+			bfddp_send_control_packet(bs, arg);
+		break;
+	case STATE_INIT:
+		bs->bs_state = STATE_UP;
+
+		/* Start polling. */
+		bs->bs_poll = true;
+
+		/* Notify state change. */
+		bfddp_send_session_state_change(bs);
+		break;
+	case STATE_UP:
+		/* NOTHING: we haven't and the peer hasn't sent INIT yet. */
+		break;
+	}
```

Comment:
Consider adding a `default:` with `assert(0)` and immediate exit, to catch memory errors

Reply:
This is a library, I'm not sure it'd be wise to add an assert here. If we were to add an `assert` I think it would be best to add it on "entrance" functions instead in the middle of the way, there is no way to properly return error codes here.

Reply:
`assert()`s are perfectly fine in a library, when used properly - read here, a bug in the code. They shall _never_ be used when validating input from outside the program (read, input from a user, network, file etc..), as these errors are not bugs for the library.

In this case, as I understand it, if a value is read outside of the `case` list, then surely there is a bug; either memory corruption or an unhandled, possibly new, enum.

As you know `assert()` will not crash when not compiled with debug.

I would suggest to add a `default:` that calls `assert()`  (or better, wraps assert to log the critical condition when not compiled with `DEBUG`), and return from the function with an error code, or at least with a consistent state, so that when 
 execution can be attempted to be continued when compiled without `DEBUG`.

Reply:
> assert()s are perfectly fine in a library, when used properly - read here, a bug in the code. They shall never be used when validating input from outside the program (read, input from a user, network, file etc..), as these errors are not bugs for the library.

I agree.

> In this case, as I understand it, if a value is read outside of the case list, then surely there is a bug; either memory corruption or an unhandled, possibly new, enum.

This is handled with the warning: `-Wswitch-enum` (see `CMakeLists.txt`)

> I would suggest to add a default: that calls assert() (or better, wraps assert to log the critical condition when not compiled with DEBUG), and return from the function with an error code, or at least with a consistent state, so that when
execution can be attempted to be continued when compiled without DEBUG.

Makes sense, I'll try to fix this.